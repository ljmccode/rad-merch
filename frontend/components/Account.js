import { useUser } from './User';
import styled from 'styled-components';

export default function Account() {
  const user = useUser();

  const AccountStyles = styled.div`
    margin: 0 auto;
    margin-top: 3rem;
    border: 1px solid var(--black);
    border-radius: 5px;
    padding: 2rem;
    padding-left: 4rem;
    line-height: 0.5;
    box-shadow: var(--bs);
    max-width: 800px;

    .name {
      margin-bottom: 4rem;
    }

    .fields {
      color: var(--dark-orange);
      font-size: 2rem;
    }
  `;

  return (
    <AccountStyles>
      <div className="name">
        <p className="fields">Name:</p>
        <p>{user?.name}</p>
      </div>
      <div>
        <p className="fields">Email:</p>
        <p>{user?.email}</p>
      </div>
    </AccountStyles>
  );
}
