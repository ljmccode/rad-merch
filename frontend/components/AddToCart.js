/* eslint-disable prettier/prettier */
import { useMutation } from '@apollo/client';
import gql from 'graphql-tag';
import { string } from 'prop-types';
import { CURRENT_USER_QUERY } from './User';

export const ADD_TO_CART_MUTATION = gql`
  mutation ADD_TO_CART_MUTATION($id: ID!) {
    addToCart(productId: $id) {
      id
    }
  }
`;

export default function AddToCart({ id, user }) {
  const [addToCart, { loading }] = useMutation(ADD_TO_CART_MUTATION, {
    variables: { id },
    refetchQueries: [{ query: CURRENT_USER_QUERY }],
  });

  async function handleSubmit(e) {
    e.preventDefault();
    !user ? alert('Please sign in to add to cart!') : await addToCart();
  }

  return (
    <button disabled={loading} type="button" onClick={handleSubmit}>
      Add{loading && 'ing'} To Cart
    </button>
  );
}

AddToCart.propTypes = {
  id: string,
};
