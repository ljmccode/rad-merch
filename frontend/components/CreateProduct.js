import { useMutation } from '@apollo/client';
import gql from 'graphql-tag';
import useForm from '../lib/useForm';
import Form from './styles/Form';

const CREATE_PRODUCT_MUTATION = gql`
  mutation CREATE_PRODUCT_MUTATION(
    # which variables are getting passed in and what types
    $name: String!
    $description: String!
    $price: Int!
    $image: Upload
  ) {
    createProduct(
      data: {
        name: $name
        description: $description
        price: $price
        status: "AVAILABLE"
        # create a relationship and an item behind the scenes
        photo: { create: { image: $image, altText: $name } }
      }
    ) {
      id
      price
      description
      name
    }
  }
`;

export default function CreateProduct() {
  const { inputs, handleChange } = useForm({
    image: '',
    name: 'Jean Jacket',
    price: 200000,
    description: 'Vintage jacket',
  });
  // first thing it returns is the function that runs the mutation
  const [createProduct, { data, error, loading }] = useMutation(
    CREATE_PRODUCT_MUTATION,
    {
      variables: inputs,
    }
  );
  console.log(createProduct);
  return (
    <Form
      onSubmit={async (e) => {
        e.preventDefault();
        console.log(inputs);
        // Submit the input fields to the backend
        const res = await createProduct();
        console.log(res);
      }}
    >
      <fieldset>
        <label htmlFor="image">
          Image
          <input
            required
            type="file"
            id="image"
            name="image"
            onChange={handleChange}
          />
        </label>
        <label htmlFor="name">
          Name
          <input
            type="text"
            id="name"
            name="name"
            placeholder="Name"
            value={inputs.name}
            onChange={handleChange}
          />
        </label>
        <label htmlFor="price">
          Price
          <input
            type="number"
            id="price"
            name="price"
            placeholder="Price"
            value={inputs.price}
            onChange={handleChange}
          />
        </label>
        <label htmlFor="description">
          Description
          <textarea
            id="description"
            name="description"
            placeholder="Description"
            value={inputs.description}
            onChange={handleChange}
          />
        </label>

        <button type="submit">Add Product</button>
      </fieldset>
    </Form>
  );
}
