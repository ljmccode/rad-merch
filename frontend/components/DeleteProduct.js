import { useMutation } from '@apollo/client';
import gql from 'graphql-tag';

const DELETE_PRODUCT_MUTATION = gql`
  mutation DELETE_PRODUCT_MUTATION($id: ID!) {
    deleteProduct(id: $id) {
      id
      name
    }
  }
`;

function update(cache, payload) {
  // within payload.data.deleteProduct is the id and name that we asked to get returned
  // cache.evict removes the prduct with the given id from the cache
  cache.evict(cache.identify(payload.data.deleteProduct));
}

export default function DeleteProduct({ id, children }) {
  // deleteProduct is the mutation function
  const [deleteProduct, { loading }] = useMutation(DELETE_PRODUCT_MUTATION, {
    variables: { id },
    update,
  });
  return (
    <button
      type="button"
      disabled={loading}
      onClick={() => {
        if (confirm('Are you sure you want to delete this item')) {
          console.log('deleting');
          deleteProduct().catch((err) => alert(err.message));
        }
      }}
    >
      {children}
    </button>
  );
}
