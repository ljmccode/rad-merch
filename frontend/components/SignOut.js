import { useMutation } from '@apollo/client';
import gql from 'graphql-tag';
import Router from 'next/router';
import { CURRENT_USER_QUERY } from './User';

/* eslint-disable prettier/prettier */
const SIGN_OUT_MUTATION = gql`
  mutation {
    endSession
  }
`;

export default function SignOut() {
  const [signout] = useMutation(SIGN_OUT_MUTATION, {
    refetchQueries: [{ query: CURRENT_USER_QUERY }],
  });

  async function handleSubmit(e) {
    e.preventDefault();
    await signout().catch(console.error);
    Router.push({
      pathname: `/products`,
    });
  }

  return (
    <button type="button" onClick={handleSubmit}>
      Sign Out
    </button>
  );
}
