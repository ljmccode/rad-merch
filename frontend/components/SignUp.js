/* eslint-disable prettier/prettier */
import { useMutation } from '@apollo/client';
import gql from 'graphql-tag';
import useForm from '../lib/useForm';
import Form from './styles/Form';
import Error from './ErrorMessage';

export const SIGNUP_MUTATION = gql`
  mutation SIGNUP_MUTATION(
    $email: String!
    $name: String!
    $password: String!
  ) {
    createUser(data: { email: $email, name: $name, password: $password }) {
      id
      email
      name
    }
  }
`;

export default function SignUp() {
  const { inputs, handleChange, resetForm } = useForm({
    email: '',
    name: '',
    password: '',
  });

  const [signup, { data, loading, error }] = useMutation(SIGNUP_MUTATION, {
    variables: inputs,
  });

  async function handleSubmit(e) {
    e.preventDefault();
    await signup();
    resetForm();
  }

  return (
    <Form method="POST" onSubmit={handleSubmit}>
      <h2>Sign Up For an Account</h2>
      <Error error={error} />
      <fieldset>
        {data?.createUser && (
          <p>Signed up with {data.createUser.email} - You Can Now Sign In!</p>
        )}
        <label htmlFor="name">
          Your Name
          <input
            type="name"
            name="name"
            placeholder="Your Name"
            autoComplete="name"
            value={inputs.name}
            onChange={handleChange}
          />
        </label>
        <label htmlFor="email">
          Email
          <input
            type="email"
            name="email"
            placeholder="Your Email Address"
            autoComplete="email"
            value={inputs.email}
            onChange={handleChange}
          />
        </label>
        <label htmlFor="passwrod">
          Password
          <input
            type="password"
            name="password"
            placeholder="Password"
            autoComplete="password"
            value={inputs.password}
            onChange={handleChange}
          />
        </label>
        <button type="submit">Sign Up!</button>
      </fieldset>
    </Form>
  );
}
