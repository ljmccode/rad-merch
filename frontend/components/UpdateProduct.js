import { useMutation, useQuery } from '@apollo/client';
import gql from 'graphql-tag';

const SINGLE_PRODUCT_QUERY = gql`
  query SINGLE_PRODUCT_QUERY($id: ID!) {
    Product(where: { id: $id }) {
      id
      name
      description
      price
    }
  }
`;

const UPDATE_PRODUCT_MUTATION = gql`
  mutation UPDATE_PRODUCT_MUTATION(
    $id: ID!
    $name: String
    $description: String
    $price: Int
  ) {
    updateProduct(
      id: $id,
      data {
        id: $id
        name: $name
        description: $description
        price: $price 
      }
    ) {
      id
      name
      description
      price
    }
  }
  `;

export default function UpdateProduct({ id }) {
  const { data, error, loading } = useQuery(SINGLE_PRODUCT_QUERY, {
    variables: { id },
  });

  const [
    updateProduct,
    { data: updateData, error: updateError, loading: updateLoading },
  ] = useMutation(UPDATE_PRODUCT_MUTATION, {
    variables: {
      id,
      // TODO: Pass in updates to product
    },
  });

  return <p>Update {id}</p>;
}
