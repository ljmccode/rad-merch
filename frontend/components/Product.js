/* eslint-disable prettier/prettier */
import Link from 'next/link';
import PropTypes from 'prop-types';
import ItemStyles from './styles/ItemStyles';
import Title from './styles/Title';
import PriceTag from './styles/PriceTag';
import formatMoney from '../lib/formatMoney';
import DeleteProduct from './DeleteProduct';
import AddToCart from './AddToCart';
import { useUser } from './User';

export default function Product({ product }) {
  const me = useUser();

  function belongsToUser() {
    return me.products.find((userProduct) => userProduct.id === product.id);
  }

  return (
    <ItemStyles>
      <img
        src={product?.photo?.image?.publicUrlTransformed}
        alt={product.name}
      />
      <Title>
        <Link href={`/product/${product.id}`}>{product.name}</Link>
      </Title>
      <PriceTag>{formatMoney(product.price)}</PriceTag>
      <p>{product.description}</p>
      <div className="buttonList">
        <AddToCart id={product.id} user={me} />
        {me && belongsToUser() && (
          <>
            <Link
              href={{
                pathname: '/update',
                query: {
                  id: product.id,
                },
              }}
            >
              Edit
            </Link>
            <DeleteProduct id={product.id}>Delete</DeleteProduct>
          </>
        )}
      </div>
    </ItemStyles>
  );
}

Product.propTypes = {
  product: PropTypes.object,
};
