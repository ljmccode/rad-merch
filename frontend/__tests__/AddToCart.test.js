import { render, screen, waitFor } from '@testing-library/react';
import { MockedProvider } from '@apollo/react-testing';
import userEvent from '@testing-library/user-event';
import wait from 'waait';
import AddToCart from '../components/AddToCart';
import { ADD_TO_CART_MUTATION } from '../components/AddToCart';
import { CURRENT_USER_QUERY } from '../components/User';
import { fakeCartItem, fakeItem, fakeUser } from '../lib/testUtils';

const product = fakeItem();
const me = fakeUser();
const cartItem = fakeCartItem();

const mocks = [
  {
    request: {
      query: ADD_TO_CART_MUTATION,
      variables: { id: product.id },
    },
    result: {
      data: {
        addToCart: cartItem,
      },
    },
  },
  {
    request: { query: CURRENT_USER_QUERY },
    result: {
      data: {
        authenticatedItem: me,
      },
    },
  },
];

describe('<AddToCart />', () => {
  it('renders and matched snapshot', () => {
    const { container } = render(
      <MockedProvider>
        <AddToCart id={'abc123'} user={me} />
      </MockedProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it('disables add to cart button and updates text while loading', async () => {
    const { container, debug } = render(
      <MockedProvider mocks={mocks}>
        <AddToCart id="abc123" user={me} />
      </MockedProvider>
    );
    await userEvent.click(screen.getByText(/Add To Cart/i));
    const disabled = screen.getByText(/Adding To Cart/i);
    expect(disabled).toBeInTheDocument();
    expect(screen.getByText(/Adding To Cart/i)).toHaveAttribute('disabled');
  });

  it('returns to not being disabled after mutation', async () => {
    const { container, debug } = render(
      <MockedProvider mocks={mocks}>
        <AddToCart id="abc123" user={me} />
      </MockedProvider>
    );
    await userEvent.click(screen.getByText(/Add To Cart/i));
    await waitFor(() => wait(400));
    expect(screen.getByText(/Add To Cart/i)).toBeInTheDocument();
  });
});
