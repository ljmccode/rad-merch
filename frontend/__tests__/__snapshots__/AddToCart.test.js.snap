// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddToCart /> renders and matched snapshot 1`] = `
<div>
  <button
    type="button"
  >
    Add
     To Cart
  </button>
</div>
`;
